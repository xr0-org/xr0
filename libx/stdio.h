#ifndef STDIO_H
#define STDIO_H

#include <stddef.h>

typedef int FILE;

axiom int
putc(int c);

axiom int
putchar(int c);

axiom int
fputs(char *s, FILE *stream);

axiom int
puts(char *s);

axiom FILE *
fopen(char *pathname, char *mode);

axiom int
fseek(FILE *stream, int offset, int whence); /* TODO: convert offset to long */

#define SEEK_END 0
#define SEEK_SET 0

axiom int
ftell(FILE *stream);

axiom size_t
fread(void *ptr, size_t size, size_t nmemb, FILE *stream);

axiom int
fclose(FILE *stream);

/* XXX: provisionally rigid for preconditions example */
axiom char *
sprintf(char *buf, char *format, int *arg) ~ [
	*buf = $;
];

/* XXX: provisionally rigid for preconditions example */
axiom int
scanf(char *format, int *arg) ~ [
	/* since there's a sideeffect here for arg, we need to characterise it */
	*arg = $;
];

#endif
