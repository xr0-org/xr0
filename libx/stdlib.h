#ifndef STDLIB_H
#define STDLIB_H

axiom void *
malloc(int size) [ .alloc result; ];

axiom void
free(void *ptr) [ .dealloc ptr; ];

#endif
