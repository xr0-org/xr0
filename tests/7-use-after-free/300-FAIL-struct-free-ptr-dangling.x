#include <stdlib.h>
#include <stdio.h>

struct report {
	int n;
	struct score **scores;
};

struct score {
	char *subject;
	int grade;
};

struct score *
create_score(char *subject, int grade) ~ [
	pre: .alloc subject;
	.alloc result;
	result->subject = subject;
	result->grade = grade;
] {
	struct score *s;
	s = malloc(sizeof(struct score));

	s->subject = subject;
	s->grade = grade;

	return s;
}

struct report *
create_report()
{
	struct score *s;
	char *sub;

	sub = malloc(sizeof(char) * 7);
	*sub = "english";

	s = create_score(sub, 1);
	free(sub);		/* dangling ptr */
	puts(s->subject);
}
