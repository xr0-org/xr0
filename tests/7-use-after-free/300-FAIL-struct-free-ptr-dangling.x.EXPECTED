7-use-after-free/300-FAIL-struct-free-ptr-dangling.x:45:18: cannot exec statement: parameter 's' must be lvalue
